intro-Cloud/ch_Terraform.tex

```diff
@@ -0,0 +1,250 @@
+}
+\end{code}
+
+And remove or comment out the rest of the lines.
+In the previous example Terraform will deploy a simple VM and use 20Gb for a persistent volume and port 56469 to connect via ssh (it also creates all required security groups).
+
+\section{Deploy Terraform templates}\label{sec:deploy-terraform-templates}
+If you have followed the previous steps now you can init and deploy your infrastucture to
+Tier-1 VSC cloud.
+
+You have to inititate Terraform first, if you didnt have deployed any template yet do this just once.
+
+Move to environment directory first:
+
+\begin{prompt}
+%\shellcmd{cd ~/openstack-templates/terraform/environment}
+\end{prompt}
+
+This command performs several different initialization steps in order to prepare the
+current working directory for use with Terraform:
+
+
+\begin{prompt}
+%\shellcmd{terraform init}
+\end{prompt}
+
+Now you can check and review your Terraform plan, from the same directory:
+
+\begin{prompt}
+%\shellcmd{terraform plan}
+\end{prompt}
+
+You will see a list of the resources required to deploy your infrastructure, Terraform also checks if there is any systax error in your templates.
+Your infrastructure is not deployed yet, review the plan and then just deploy it to VSC Tier-1 Cloud running:
+
+\begin{prompt}
+%\shellcmd{terraform apply}
+\end{prompt}
+
+Terraform will show your plan again and you will see this message:
+
+\begin{prompt}
+..
+..
+Do you want to perform these actions?
+  Terraform will perform the actions described above.
+  Only 'yes' will be accepted to approve.
+
+  Enter a value: 
+\end{prompt}
+
+Type \texttt{\emph{yes}} and press enter and wait a few seconds or minutes.
+If everything is correct and if you have enough quota Terraform will show you a message after creating all the required resources.
+
+
+\begin{prompt}
+..
+..
+module.vm_with_pf_rules_with_ssh_access.openstack_compute_instance_v2.instance_01: Still creating... [1m30s elapsed]
+module.vm_with_pf_rules_with_ssh_access.openstack_compute_instance_v2.instance_01: Creation complete after 1m35s [id=88c7d037-5c44-45b7-acce-f5e4e58b1c35]
+
+Apply complete! Resources: 4 added, 0 changed, 0 destroyed.
+\end{prompt}
+
+Your cloud infrastrucuture is ready to be used.
+
+It is important to keep a backup of your terraform directory, specially files within environment directory:
+\begin{prompt}
+%\shellcmd{~/openstack-templates/terraform/environment}
+\end{prompt}
+
+Terraform generates several files in this directory to keep track of any change in your infrastructure. If for some reason you lost or remove these files you will not able to modify or change the current Terraform plan (only directly from OpenStack).
+
+You can also modify and add more resources for the current templates. This task is out of the scope of this document, please refer to official Terraform documentation to add you own changes \url{https://www.terraform.io/docs} or ask to VSC Cloud admins via email at \cloudinfo.
+
```

Comment:
I successfully followed the entire manual.
A first run of 'terraform apply' yielded this error for me though:

Error: Error waiting for instance (8f1960c8-2292-4c87-bc88-662883246fd4) to become ready: unexpected state 'ERROR', wanted target 'ACTIVE'. last error: %!s(<nil>)
│ 
│   with module.vm_with_pf_rules_with_ssh_access.openstack_compute_instance_v2.instance_01,
│   on ../modules/vm_with_pf_rules_with_ssh_access/main.tf line 1, in resource "openstack_compute_instance_v2" "instance_01":
│    1: resource "openstack_compute_instance_v2" "instance_01" {
│ 
╵
╷
│ Error: Error waiting for instance (8834a023-4068-4ef6-851e-08a05fe1392a) to become ready: unexpected state 'ERROR', wanted target 'ACTIVE'. last error: %!s(<nil>)
│ 
│   with module.vm_with_pf_rules_with_ssh_access_with_nginx.openstack_compute_instance_v2.instance_01,
│   on ../modules/vm_with_pf_rules_with_ssh_access_with_nginx/main.tf line 8, in resource "openstack_compute_instance_v2" "instance_01":
│    8: resource "openstack_compute_instance_v2" "instance_01" {
│ 

Second attempt yielded 'Apply complete'